File: src/sim.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable, List, Optional, Tuple
import heapq
import math

# discrete event simulation - runs callbacks at scheduled times

@dataclass(order=True)
class Event:
    time: float  # when this event fires
    seq: int  # breaks ties so events fire in order added
    callback: Callable[..., Any]  # what to run
    args: Tuple[Any, ...]
    kwargs: dict


class Simulation:
    def __init__(self, end_time: float = math.inf) -> None:
        self._queue: List[Event] = []  # min heap of events
        self._seq = 0  # counter for event ordering
        self._time = 0.0  # current sim time
        self._end_time = end_time

    @property
    def current_time(self) -> float:
        return self._time

    def schedule_event(self, time: float, callback: Callable[..., Any], *args: Any, **kwargs: Any) -> None:
        # add an event to the queue
        if time < self._time:
            raise ValueError("Cannot schedule event in the past")
        event = Event(time, self._seq, callback, args, kwargs)
        self._seq += 1
        heapq.heappush(self._queue, event)

    def run_step(self) -> Optional[Event]:
        # run the next event in the queue
        if not self._queue:
            return None
        event = heapq.heappop(self._queue)
        if event.time > self._end_time:
            heapq.heappush(self._queue, event)  # put it back
            return None
        self._time = event.time  # advance sim time
        event.callback(*event.args, **event.kwargs)  # fire the event
        return event

    def run_until(self, end_time: Optional[float] = None) -> None:
        # run simulation until end time or queue empty
        if end_time is not None:
            self._end_time = end_time
        while self._queue and self._time <= self._end_time:
            if self.run_step() is None:
                break

    def is_done(self) -> bool:
        return not self._queue or self._time >= self._end_time


def init_sim(end_time: float = math.inf) -> Simulation:
    return Simulation(end_time=end_time)

File: src.rs/src/sim.rs
use std::cmp::Ordering;
use std::collections::BinaryHeap;

pub struct Event {
    pub time: f64,
    pub seq: u64,
    pub callback: Box<dyn FnMut()>,
}

impl Eq for Event {}

impl PartialEq for Event {
    fn eq(&self, other: &Self) -> bool {
        self.time == other.time && self.seq == other.seq
    }
}

impl Ord for Event {
    fn cmp(&self, other: &Self) -> Ordering {
        other
            .time
            .total_cmp(&self.time)
            .then_with(|| other.seq.cmp(&self.seq))
    }
}

impl PartialOrd for Event {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

pub struct Simulation {
    queue: BinaryHeap<Event>,
    seq: u64,
    time: f64,
    end_time: f64,
}

impl Simulation {
    pub fn new(end_time: f64) -> Self {
        Self {
            queue: BinaryHeap::new(),
            seq: 0,
            time: 0.0,
            end_time,
        }
    }

    pub fn current_time(&self) -> f64 {
        self.time
    }

    pub fn schedule_event<F>(&mut self, time: f64, callback: F)
    where
        F: FnMut() + 'static,
    {
        if time < self.time {
            return;
        }
        let event = Event {
            time,
            seq: self.seq,
            callback: Box::new(callback),
        };
        self.seq += 1;
        self.queue.push(event);
    }

    pub fn run_step(&mut self) -> Option<f64> {
        if let Some(mut event) = self.queue.pop() {
            if event.time > self.end_time {
                self.queue.push(event);
                return None;
            }
            self.time = event.time;
            (event.callback)();
            return Some(self.time);
        }
        None
    }

    pub fn run_until(&mut self, end_time: f64) {
        self.end_time = end_time;
        while !self.queue.is_empty() {
            if self.run_step().is_none() {
                break;
            }
        }
    }

    pub fn is_done(&self) -> bool {
        self.queue.is_empty() || self.time >= self.end_time
    }
}

pub fn init_sim(end_time: f64) -> Simulation {
    Simulation::new(end_time)
}
